tutorial_code/embed_square_lattice.py

```diff
@@ -47,33 +45,109 @@ def embed_square_lattice(L, try_to_load=True, **kwargs):
             if x < L - 1:
                 bqm.set_quadratic(x * L + y, (x + 1) * L + y, 1)
+    return bqm
+
 
-    G = dimod.to_networkx_graph(bqm)
-    A = sampler.to_networkx_graph()
+def embed_square_lattice(
+    sampler: MockDWaveSampler,
+    L: int,
+    use_cache: bool = True,
+    **kwargs,
+) -> tuple[np.ndarray, dimod.BinaryQuadraticModel]:
+    """Embeds a square lattice of length L (LxL cylinder).
 
-    cache_filename = f"cached_embeddings/{sampler.solver.name}__L{L:02d}_square_embeddings_cached.txt"
-    if try_to_load:
+    Args:
+        sampler (int): DWaveSampler for which to embed
+        L (int): lattice length
+        use_cache (bool, default=True): When True, embeddings are
+            saved to and loaded from a local directory whenever
+            possible. If writing to a directory is not possible
+            a warning is thrown.
+
+    Returns:
+        Tuple[np.ndarray, dimod.BQM]: A matrix of embeddings and BQM for the lattice.
+    """
+
+    bqm = make_square_bqm(L)
+
+    solver_name = sampler.properties["chip_id"]  # sampler.solver.name
+    cache_filename = (
+        f"cached_embeddings/{solver_name}_L{L:02d}_square_embeddings_cached.txt"
+    )
+
+    if use_cache and os.path.exists(cache_filename):
+        embeddings = np.loadtxt(cache_filename, dtype=int)
+        if embeddings.ndim == 1:
+            embeddings = embeddings[np.newaxis, :]
+        print(f"Loaded {cache_filename}")
+        return embeddings, bqm
+    else:
+        G = dimod.to_networkx_graph(bqm)
+        A = sampler.to_networkx_graph()
+
+        if not embedding_feasibility_filter(S=G, T=A):
+            raise ValueError(f"Embedding {G} on {A} is infeasible")
+        sublattice_size = kwargs.pop(
+            "sublattice_size",
+            min(lattice_size_lower_bound(S=G, T=A) + 1, max(A.graph.get("rows"), A.graph.get("columns"))),
+        )
+        if not isinstance(sublattice_size, int) or sublattice_size <= 0:
+            raise ValueError(
+                f"'sublattice_size' must be a positive integer. Received {sublattice_size}."
+            )
+
+        print(
+            "Creating embeddings may take several minutes."
+            "\nTo accelerate the process a smaller lattice (L) might be "
+            "considered and/or the search restricted to max_num_emb=1."
+        )
+        max_num_emb = kwargs.pop("max_num_emb", 1)  # many can be slow
+        embedder_kwargs = {"timeout": kwargs.pop("timeout", 10)}
+        embeddings = embeddings_to_array(
+            find_sublattice_embeddings(
+                S=G,
+                T=A,
+                sublattice_size=sublattice_size,
+                max_num_emb=max_num_emb,
+                embedder_kwargs=embedder_kwargs,
+                **kwargs,
+            ),
+            node_order=sorted(G.nodes()),
+            as_ndarray=True,
+        )
+
+        if embeddings.size == 0:
+            raise ValueError(
```

Comment:
Similar to comment above, perhaps use a custom exception here?

Reply:
Too lazy, RunTimeError then, or TooLazyError?

Reply:
`RunTimeError` is fine then. It's sort-of a catch all for errors that don't fall into any other categories.